src/Microsoft.AspNetCore.ResponseCaching/ResponseCacheMiddleware.cs

```diff
@@ -355,7 +355,7 @@ internal static bool ConditionalRequestSatisfied(ResponseCacheContext context)
                     }
                 }
             }
-            else if ((cachedResponseHeaders.LastModified ?? cachedResponseHeaders.Date) <= context.TypedRequestHeaders.IfUnmodifiedSince)
+            else if (context.TypedRequestHeaders.IfUnmodifiedSince != null && (cachedResponseHeaders.LastModified ?? cachedResponseHeaders.Date) <= context.TypedRequestHeaders.IfUnmodifiedSince)
```

Comment:
Remember that ResponseHeaders properties have no caching, they parse every time you access them